language/move-stdlib/nursery/Vault.move

```diff
@@ -0,0 +1,537 @@
+        &accessor.content
+    }
+
+    /// Releases read accessor.
+    public fun release_read_accessor<Content: store + drop>(accessor: ReadAccessor<Content>)
+    acquires Vault {
+        let ReadAccessor{ content: new_content, vault_address } = accessor;
+        let content = &mut borrow_global_mut<Vault<Content>>(vault_address).content;
+        // We (should be/are) able to prove that the below cannot happen, but we leave the assertion
+        // here anyway for double safety.
+        assert(Option::is_none(content), Errors::internal(EACCESSOR_INCONSISTENCY));
+        Option::fill(content, new_content);
+    }
+
+    // ================================================================================================================
+    // Modify Accessor
+
+    /// A modify accessor for the content of the vault.
+    struct ModifyAccessor<Content: store + drop> {
+        content: Content,
+        vault_address: address,
+    }
+
+    /// Creates a modify accessor for the content in the vault based on a modify capability. This
+    /// is similar like `Self::read_accessor` but the returned accessor will allow to mutate
+    /// the content.
+    public fun modify_accessor<Content: store + drop>(cap: &ModifyCap<Content>): ModifyAccessor<Content>
+    acquires Vault {
+        let content = &mut borrow_global_mut<Vault<Content>>(cap.vault_address).content;
+        assert(Option::is_some(content), Errors::invalid_state(EACCESSOR_IN_USE));
+        ModifyAccessor{ vault_address: cap.vault_address, content: Option::extract(content) }
+    }
+
+    /// Returns a mutable reference to the content represented by a modify accessor.
+    public fun borrow_mut<Content: store + drop>(accessor: &mut ModifyAccessor<Content>): &mut Content {
+        &mut accessor.content
+    }
+
+    /// Releases a modify accessor. This will ensure that any modifications are written back
+    /// to the vault.
+    public fun release_modify_accessor<Content: store + drop>(accessor: ModifyAccessor<Content>)
+    acquires Vault {
+        let ModifyAccessor{ content: new_content, vault_address } = accessor;
+        let content = &mut borrow_global_mut<Vault<Content>>(vault_address).content;
+        // We (should be/are) able to prove that the below cannot happen, but we leave the assertion
+        // here anyway for double safety.
+        assert(Option::is_none(content), Errors::internal(EACCESSOR_INCONSISTENCY));
+        Option::fill(content, new_content);
+    }
+
+
+    // ================================================================================================================
+    // Delegation
+
+    /// Delegates the right to acquire a capability of the given type. Delegation must have been enabled
+    /// during vault creation for this to succeed.
+    public fun delegate<Content: store + drop>(cap: &DelegateCap<Content>, to_signer: &signer, cap_type: CapType)
```

Comment:
It seems that "delegation" can be done only in multi-agent transactions. Both of the one with the delegate cap and the new delegate (i.e., `to_signer`) need to be signers.

Reply:
Yes, and this is essential. Every authorization needs to be signer based, otherwise it is pretty useless for access control, because a simple address can be faked.